workflow/controller/operator.go

```diff
@@ -136,14 +171,16 @@ func (wfc *WorkflowController) operateWorkflow(wf *wfv1.Workflow) {
 		return
 	}
 
-	// TODO: workflow finalizer logic goes here
-
-	// If we get here, the workflow completed, all PVCs were deleted successfully,
-	// and finalizers were executed (finalizer feature yet to be implemented).
-	// We now need to infer the workflow phase from the node phase.
+	// If we get here, the workflow completed, all PVCs were deleted successfully, and
+	// exit handlers were executed. We now need to infer the workflow phase from the
+	// node phase.
 	switch node.Phase {
 	case wfv1.NodeSucceeded, wfv1.NodeSkipped:
-		woc.markWorkflowSuccess()
+		if onExitNode != nil && !onExitNode.Successful() {
+			woc.markWorkflowSuccess()
```

Comment:
`woc.markWorkflowSuccess()` is called in both cases. Is that intentional?

Reply:
This is a bug. This should be markWorkflowFailed. The logic is supposed to be: if the main workflow succeeded, but the onExitNode failed, we consider the workflow as a whole failed. Thanks for catching this!